In [86]:
import pandas as pd
import numpy as np

In [87]:
df = pd.read_csv('apple_quality.csv')

In [88]:
df.head()

A_id      Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0   0.0 -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840   
1   1.0 -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530   
2   2.0 -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033   
3   3.0 -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761   
4   4.0  1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849   

        Acidity Quality  
0  -0.491590483    good  
1  -0.722809367    good  
2   2.621636473     bad  
3   0.790723217    good  
4   0.501984036    good

In [89]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4001 entries, 0 to 4000
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   A_id         4000 non-null   float64
 1   Size         4000 non-null   float64
 2   Weight       4000 non-null   float64
 3   Sweetness    4000 non-null   float64
 4   Crunchiness  4000 non-null   float64
 5   Juiciness    4000 non-null   float64
 6   Ripeness     4000 non-null   float64
 7   Acidity      4001 non-null   object 
 8   Quality      4000 non-null   object 
dtypes: float64(7), object(2)
memory usage: 281.4+ KB


In [90]:
df = df.drop('A_id',axis=1)

In [91]:
df.head()

Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0 -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840   
1 -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530   
2 -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033   
3 -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761   
4  1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849   

        Acidity Quality  
0  -0.491590483    good  
1  -0.722809367    good  
2   2.621636473     bad  
3   0.790723217    good  
4   0.501984036    good

In [92]:
df.isnull().sum()

Size           1
Weight         1
Sweetness      1
Crunchiness    1
Juiciness      1
Ripeness       1
Acidity        0
Quality        1
dtype: int64

In [93]:
df = df.dropna()

In [94]:
df.isnull().sum()

Size           0
Weight         0
Sweetness      0
Crunchiness    0
Juiciness      0
Ripeness       0
Acidity        0
Quality        0
dtype: int64

In [95]:
df.shape

(4000, 8)

In [96]:
df['Acidity'] = df['Acidity'].astype('float64')

In [97]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4000 entries, 0 to 3999
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Size         4000 non-null   float64
 1   Weight       4000 non-null   float64
 2   Sweetness    4000 non-null   float64
 3   Crunchiness  4000 non-null   float64
 4   Juiciness    4000 non-null   float64
 5   Ripeness     4000 non-null   float64
 6   Acidity      4000 non-null   float64
 7   Quality      4000 non-null   object 
dtypes: float64(7), object(1)
memory usage: 281.2+ KB


In [98]:
from sklearn.preprocessing import LabelEncoder

#Usaremos essa lib para poder utilizar a variavel 'Quality' no modelo, alterando seu tipo

In [99]:
label_encoder = LabelEncoder()
df['Quality'] = label_encoder.fit_transform(df['Quality'])

In [100]:
df['Quality']

0       1
1       1
2       0
3       1
4       1
       ..
3995    0
3996    1
3997    0
3998    1
3999    1
Name: Quality, Length: 4000, dtype: int32

# Modelando e padronizando o dataframe para contrução do modelo

In [101]:
#Padronização dos dados

from sklearn.preprocessing import StandardScaler

#Ao aplicar o StandardScaler durante o pré-processamento, garantimos que os dados estejam em uma escala uniforme, facilitando o treinamento do modelo.

In [102]:
scaler = StandardScaler()
df[['Size','Weight','Sweetness','Crunchiness','Juiciness','Ripeness','Acidity']] = scaler.fit_transform(df[['Size','Weight','Sweetness','Crunchiness','Juiciness','Ripeness','Acidity']])
#Todos os valores  exceto os valores de queremos predizer que no caso é o 'Quality'

In [103]:
df

Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0    -1.798424 -0.950373   2.993421    -1.424150   0.690545 -0.089872   
1    -0.359060 -1.154404   2.127698     0.429746   0.176767  0.197020   
2     0.109445 -0.225759  -0.652507    -0.946892   1.205422 -0.286156   
3    -0.079977 -0.800146   0.923916    -0.772399   1.619575 -2.087320   
4     0.968573 -0.191640   0.044164    -1.096894   1.305025 -0.961548   
...        ...       ...        ...          ...        ...       ...   
3995  0.291729 -0.048594  -1.669449    -0.365345   0.614425  0.931482   
3996  0.108878  1.834105   0.137124    -1.159058  -0.252634 -0.846326   
3997 -1.105655 -0.716904  -1.013784    -0.234036   0.874379  2.275957   
3998 -1.818112 -0.492908   1.459901    -0.845446   0.854549 -0.151419   
3999  0.405409 -0.453071   0.304496    -1.525439   0.390954 -0.680212   

       Acidity  Quality  
0    -0.269415        1  
1    -0.378997        1  
2     1.206044        0  
3     0.338315        1  
4     0.201472        1  
...        ...      ...  
3995  0.028866        0  
3996  0.842347        1  
3997 -0.668950        0  
3998 -1.093171        1  
3999  0.721761        1  

[4000 rows x 8 columns]

# Criando do modelo

In [104]:
#Divisao entre treino e teste

X = df.drop('Quality',axis=1)
y = df['Quality']

In [105]:
#importação das bibliotecas 

from sklearn.model_selection import train_test_split
#Divide um conjunto de dados entre treino e teste para avaliar modelos de machine learning

from sklearn.linear_model import LogisticRegression
#implementa regressao logistica para problemas de classificacao binaria.estima probabilidade de pertencimento de classe

from sklearn.tree import DecisionTreeClassifier
#Cria uma arvore de decisao para cada classificação, divide dados baseado em regras de decisao em cada nó

from sklearn.ensemble import RandomForestClassifier
#Combina varias arvoes de decisao para melhorar a precisao e reduzir o overfitting, bom para classificação e regressão

from sklearn.svm import SVC
#implementa maquinas de vetores de suporte para classificação, encontra hiperplano que melhor separa classes

from sklearn.neighbors import KNeighborsClassifier
#Classifica as instancias com base nas classes mais proximas (vizinhos)

#from xgboost import XGBClassifier
#Implementa algoritmo de GRADIENT BOOSTING, bom para classificacao e regressao. EFICIENTE-ESCALONAVEL COM REGULARIZAÇÃO

from sklearn.metrics import accuracy_score, f1_score, classification_report
#ACCURACY -> é uma métrica comum para avaliar o desempenho de um modelo de classificação, mede proporção de previsoes corretas
#F1_SCORE -> é uma métrica que combina precisão com recall, util em classes desequilibradas. calcula média harmonica entre precisão e recall
#CLASSIFICATION_REPORT -> visão geral das métricas de avaliação

from sklearn.neural_network import MLPClassifier
#implementa redes neurais para aprender relações complexas entre recursos e classes

In [106]:
X_train,X_test,y_train,y_test = train_test_split(X,
                                                y,
                                                test_size=0.2,
                                                random_state=42)

In [107]:
#Lista de classificações

classifiers = {
    'Logistic Regression': LogisticRegression(),
    'Decision Tree': DecisionTreeClassifier(),
    'Random Forest': RandomForestClassifier(),
    'SVC': SVC(),
    'K-nearest neigbour': KNeighborsClassifier(),
    'Neural Network': MLPClassifier(),    
}

In [108]:
accuracies = []
f1_scores = []

In [109]:
for name, clf in classifiers.items():
    clf.fit(X_train,y_train)
    y_pred = clf.predict(X_test)
    
    accuracy = accuracy_score(y_test,y_pred)
    
    accuracies.append(accuracy)
    
    print(f"\n {30*'='}\n{name}\n{30*' '}")
    print(f"Accuracy: {accuracy:.4f}")
    print(f"Classification Report")
    print(classification_report(y_test,y_pred))


Logistic Regression
                              
Accuracy: 0.7538
Classification Report
              precision    recall  f1-score   support

           0       0.75      0.76      0.75       401
           1       0.75      0.75      0.75       399

    accuracy                           0.75       800
   macro avg       0.75      0.75      0.75       800
weighted avg       0.75      0.75      0.75       800


Decision Tree
                              
Accuracy: 0.8075
Classification Report
              precision    recall  f1-score   support

           0       0.81      0.81      0.81       401
           1       0.81      0.81      0.81       399

    accuracy                           0.81       800
   macro avg       0.81      0.81      0.81       800
weighted avg       0.81      0.81      0.81       800


Random Forest
                              
Accuracy: 0.9050
Classification Report
              precision    recall  f1-score   support

           0       0.91      0

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


A acurácia 
- Mede a precisão e exatidão dos resultados obtidos. Mede o quão próximos estão os dados em uma linha de regressão ajustada.

Precisão(Precision)
- Mede a precisão de verdadeiros positivos, útil para filtrar falsos positivos

Revocação(Recall)
- Parecido com o Precision, mas sua utilidade é voltada para evitar falsos negativos

F1-Score
- É uma média harmônica entre a Precision e o Recall, equilibra a importância de ambas métricas sendo útil em classes desequilibradas 
- 0-0.5 RUIM
- 0.5-0.8 ACEITAVEL
- 0.8-0.9 BOM
- 0.9-1.0 MUITO BOM

Support
- Valor total de instâncias na classe


# 1. Logistic Regression

Método usado para ajustar modelo de regressão quando a variável resposta é binária (no caso a 'Quality')

Acurácia 75.38% // F1 Score 0.75

O modelo acertou aproximandamente 75.38% das previsões (BOM SCORE)


# 2. Decision Tree

Funciona como um tipo de fluxograma, lida com dados complexos e pode ser 'podada' para evitar overfitting

Acurácia 80.87% // F1 Score 0.81

A Classificação Decision Tree teve uma boa acurácia e uma boa performance entre as classes, Bom resultado!

# 3. Random Forest 

A Random Forest combina a previsão de várias árvores por meio de médias, melhorando a precisão e controlando o overfitting. Sistema bem robusto.

Acurácia 90.5% // F1 Score 0.91

Random Forest até o momento nos apresentou os melhores resultados, indicando capacidade de aprender a complexidade do dataset

## 4. SVM

O SVM é uma ferramenta poderosa para resolver problemas de classificação e regressão, encontrando hiperplanos que separam os dados de maneira eficaz

Acurácia 90.38% // F1 Score 0.90

O modelo SVM apresentou alto score também, indicando eficácia no tratamento de relações não lineares

# 5. K-nearest Neighbors

O K-nearest Neighbors ou 'KNN' é um classificador de aprendizado supervisionado não paramétrico que utiliza a proximidade para fazer classificações ou previsões sobre o agrupamento de um ponto de dados individual

Acurácia 90.12% // F1 Score 0.90

Também demonstra uma alta performance como os outros modelos

# Neural Network

A Neural Network  são essenciais para resolver problemas de classificação; previsão e modemlagem complexa. Seu funcionamento é inspirado na comunicação entre os neurônios do cérebro humano

Acurácia 92.25% // F1 Score 0.92

Modelo com a acurácia mais alta